beps/0005-definitive-plugin-listing/README.md

```diff
@@ -0,0 +1,250 @@
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this BEP. Describe why the change is important and the benefits to users.
+-->
+
+Split backends are a consistently difficult space to operate in and design for. There has been a growing desire for the framework to provide a way to get a list of the installed plugins. This was nearly impossible in the old backend, where plugins were hosted on denormalized routes and had non-standard startup sequences. In the new backend, this has become significantly more doable. Moving this forward would unblock a number of cases that require knowledge of your entire Backstage installation, namely a single OpenAPI spec for your instance, checking installed permissions, and DevTools information.
+
+### Goals
+
+<!--
+List the specific goals of the BEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+1. As an integrator, I can now get a list of currently installed features across my deployment.
+1. As an integrator/administrator, I can get a list of currently installed features for a single instance in my deployment.
+1. As an integrator/administrator, I can continue to use my static `HostDiscovery` config.
+
+### Non-Goals
+
+<!--
+What is out of scope for this BEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+1. The list of plugins will not change unless there is a config change, they will not be dynamically loaded.
+1. The list of plugins will not be used in routing or discovery.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.
+-->
+
+Most of the work below is API definitions for hosting the list of features and better abstractions to make that list available through HTTP. This is not intended to perform much heavy lifting, that can happen in a further design. We will build a default implementation on the existing static `HostDiscovery` config definition for split deployments. Users can override the API as they need.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs or even code snippets.
+If there's any ambiguity about HOW your proposal will be implemented, this is the place to discuss them.
+-->
+
+### Glossary
+
+### New Type `BackstageFeature`
+
+```ts
+type BackstageFeature =
+  | {
+      type: 'plugin';
+      pluginId: string;
+    }
+  | {
+      type: 'module';
+      pluginId: string;
+      moduleId: string;
+    };
+```
+
+### New `InstanceMetadataService`
```

Comment:
One more thought. Making it a service implies the ability to depend on it and possibly even replace it. Does either of those two actions make any sense? Would it be an antipattern to try to enumerate features inside another plugin, while early startup is still in full swing? Essentially - what if this was dialed back to only being the `instanceInfo` HTTP endpoint and no service at all? Would that make sense?

Reply:
I think it makes sense to split into two services tbh, one that is strictly for colocated plugins and very explicit about that. That one would be provided by the backend initializer, just like the existing plugin metadata service. In addition to that we also add a separate service that is the one with more appealing naming and the one actually intended to be used for discovery. We keep that one replacable though, so that one can plugin in their own way of listing available services across all deployed instances.

Reply:
@Rugvip Splitting the `InstanceMetadata` service into 2 or using the services described here to create a `BackstageMetadataService` or something that could be used for discovery?